### ISSUE TITLE
feature: Add Band-Based Scaler and Pareto Scaler

### ISSUE THREAD

**Issue by paucablop** [paucablop/chemotools#240]
#### Summary

Introduce two additional preprocessing transformers for spectroscopic workflows:

1. **BandScaler** – scales spectra using the intensity of a user-defined spectral band.
2. **ParetoScaler** – applies standard Pareto scaling commonly used in chemometrics and metabolomics.

Both transformers would follow the standard estimator API (`fit`, `transform`) and integrate seamlessly with preprocessing pipelines.

---

# 1. BandScaler

## Motivation

In many spectroscopic workflows, spectra are normalized relative to a **reference band**. This band may correspond to:

* a stable internal standard
* a solvent peak
* a characteristic absorption region

Normalizing spectra relative to this region can reduce variability due to:

* sample concentration differences
* path length variation
* instrument fluctuations

While general-purpose normalization methods exist, a dedicated **band-based scaling transformer** would provide a flexible and transparent implementation for spectroscopic data.

---

## Proposed API

```python
BandScaler(band, method="mean", wavelengths=None)
```

Parameters:

* **band**: tuple `(start, end)`
  Spectral band used for normalization.

* **method**: `"mean"` or `"area"`

  * `"mean"` → normalize by mean intensity within the band
  * `"area"` → normalize by the area under the band

* **wavelengths**: optional array-like
  Wavelength or wavenumber axis used to determine the band indices.

---

## Behavior

Given a spectrum (x):

1. Extract values within the selected band.
2. Compute the scaling factor:

Mean-based scaling:

$s = \frac{1}{n}\sum_{i \in band} x_i$

Area-based scaling:

$s = \sum_{i \in band} x_i$

3. Normalize the spectrum:

$x' = \frac{x}{s}$

Each spectrum is scaled independently.

---

## Example Usage

```python
scaler = BandScaler(
    band=(1600, 1700),
    method="area",
    wavelengths=wavenumbers
)

X_scaled = scaler.fit_transform(X)
```

---

## Use Cases

* normalization using internal standard peaks
* reducing sample concentration variability
* spectral calibration pipelines
* spectroscopy and metabolomics preprocessing

---

# 2. ParetoScaler

## Motivation

Pareto scaling is widely used in chemometrics and metabolomics as a compromise between:

* **mean-centering**
* **autoscaling (unit variance scaling)**

Instead of dividing by the standard deviation, variables are divided by the **square root of the standard deviation**, which reduces the influence of large variance variables without amplifying noise excessively.

---

## Proposed API

```python
ParetoScaler(with_mean=True)
```

Parameters:

* **with_mean**: bool, default=True
  Whether to mean-center variables before scaling.

---

## Behavior

For each variable (j):

1. Compute mean:

$\mu_j$

2. Compute standard deviation:

$\sigma_j$

3. Apply Pareto scaling:

$x'*{ij} = \frac{x*{ij} - \mu_j}{\sqrt{\sigma_j}}$

If `with_mean=False`, only scaling is applied.

---

## Example Usage

```python
scaler = ParetoScaler()
X_scaled = scaler.fit_transform(X)
```

---

# Implementation Notes

Both scalers would follow the standard transformer pattern:

```python
fit(X, y=None)
transform(X)
fit_transform(X, y=None)
```

Attributes stored after fitting:

BandScaler:

* band indices
* wavelengths mapping (if provided)

ParetoScaler:

* variable means
* variable standard deviations

---

# Compatibility

Both transformers should integrate with preprocessing pipelines:

```python
Pipeline([
    ("band_norm", BandScaler((1500, 1600))),
    ("pareto", ParetoScaler()),
    ("model", SomeModel())
])
```

---

# Benefits

* adds domain-specific preprocessing tools for spectroscopic data
* simplifies normalization using reference bands
* introduces a commonly used chemometric scaling method
* improves reproducibility and transparency in preprocessing pipelines

---

# Potential Extensions

Future improvements could include:

* trapezoidal integration for band area estimation
* multiple-band normalization
* visualization utilities for selected bands
* optional robustness to outliers

---

If this proposal aligns with the project goals, I would be happy to contribute initial implementations of both scalers.
